code/modules/mapping/reader.dm

```diff
@@ -367,33 +423,36 @@
 		// If our relative x is greater then X upper, well then we've gotta limit our expansion
 		var/delta = max(final_x - x_delta_with, 0)
 		final_x -= delta
-	if(final_x > world.maxx && !cropMap)
+	if(final_x > world.maxx && !crop_map)
 		if(new_z)
 			// Need to avoid improperly loaded area/turf_contents
-			world.increaseMaxX(final_x, max_zs_to_load = z_offset - 1)
+			world.increaseMaxX(final_x, max_zs_to_load = z_offset - z_lower)
 		else
 			world.increaseMaxX(final_x)
 		expanded_x = TRUE
 
 	var/lowest_x = max(x_lower, 1 - x_relative_to_absolute)
 
 	// We make the assumption that the last block of turfs will have the highest embedded z in it
-	var/highest_z = last_column.zcrd + z_offset - 1 // Lets not just make a new z level each time we increment maxz
+	var/highest_z = last_column.zcrd + z_offset - z_lower // Lets not just make a new z level each time we increment maxz
 	var/z_threshold = world.maxz
-	if(highest_z > z_threshold && cropMap)
+	if(highest_z > z_threshold && crop_map)
 		for(var/i in z_threshold + 1 to highest_z) //create a new z_level if needed
 			world.incrementMaxZ()
 		if(!no_changeturf)
 			WARNING("Z-level expansion occurred without no_changeturf set, this may cause problems when /turf/AfterChange is called")
 
 	for(var/datum/grid_set/gset as anything in gridSets)
+		if(gset.zcrd > z_upper || gset.zcrd < z_lower)
+			continue // skip z levels we dont want
```

Comment:
since limiting z layer loads is an uncommon case, I think it would be better to conditionally filter gridsets if z_upper/z_lower are set to any interesting values.